scaleway/resource_compute_instance_volume.go

```diff
@@ -138,14 +168,43 @@ func resourceScalewayComputeInstanceVolumeDelete(d *schema.ResourceData, m inter
 		return err
 	}
 
-	err = instanceAPI.DeleteVolume(&instance.DeleteVolumeRequest{
-		VolumeID: id,
+	deleteRequest := &instance.DeleteVolumeRequest{
 		Zone:     zone,
-	})
-
-	if err != nil && !is404Error(err) {
-		return fmt.Errorf("couldn't delete volume: %v", err)
+		VolumeID: id,
 	}
 
-	return nil
+	err = resource.Retry(ServerRetryFuncTimeout, func() *resource.RetryError {
+		err := instanceAPI.DeleteVolume(deleteRequest)
+		if isSDKResponseError(err, http.StatusBadRequest, "a server is attached to this volume") {
+			if d.Get("type").(string) != instance.VolumeTypeBSSD.String() {
+				err = instanceAPI.ServerActionAndWait(&instance.ServerActionAndWaitRequest{
+					Zone:     zone,
+					ServerID: d.Get("server_id").(string),
+					Action:   instance.ServerActionPoweroff,
+					Timeout:  ServerWaitForTimeout,
+				})
+				if err != nil && !isSDKResponseError(err, http.StatusBadRequest, "server should be running") {
+					return resource.NonRetryableError(err)
+				}
+			}
+			_, err = instanceAPI.DetachVolume(&instance.DetachVolumeRequest{
+				Zone:     zone,
+				VolumeID: id,
+			})
+			if isSDKResponseError(err, http.StatusBadRequest, "Instance must be powered off to change local volumes") {
```

Comment:
Why would this happen if we stop the server before ?

Reply:
When 2 volumes are dettaching at the same time from the same server, one will synchronously wait for the server to be stopped, the second will receive a message "server should be running" when trying to poweroff.  Then the second will try to detach the volume too early, resulting in a "Instance must be powered off to change local volumes" error. 🤷‍♂